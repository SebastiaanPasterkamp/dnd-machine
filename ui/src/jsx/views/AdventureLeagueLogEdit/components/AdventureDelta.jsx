import React from 'react';
import PropTypes from 'prop-types';

import CalculatorInputField from '../../../components/CalculatorInputField.jsx';
import ControlGroup from '../../../components/ControlGroup.jsx';
import InputField from '../../../components/InputField.jsx';
import Panel from '../../../components/Panel.jsx';

export class AdventureDelta extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.state = {};
    }

    onChange = (earned, formula) => {
        const { starting, disabled, setState } = this.props;
        if (disabled) {
            return;
        }

        const total = starting + earned;
        this.setState(
            { formula, total },
            () => setState({ starting, earned, total })
        );
    }

    render() {
        const {
            starting, earned, total, className, label, disabled,
        } = this.props;
        const {
            formula = earned,
        } = this.state;

        return (
            <Panel
                header={label}
                className={className}
            >
                <ControlGroup label="Starting">
                    <InputField
                        placeholder="Starting..."
                        disabled={true}
                        type="number"
                        value={starting}
                    />
                </ControlGroup>
                <ControlGroup label="Earned">
                    <CalculatorInputField
                        placeholder="Earned..."
                        data-field="earned"
                        value={earned}
                        disabled={disabled}
                        setState={this.onChange}
                    />
                </ControlGroup>
                <ControlGroup label="Total">
                    <InputField
                        placeholder="Total..."
                        disabled={true}
                        type="number"
                        value={earned ? total : starting}
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureDelta.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: PropTypes.number,
    earned: PropTypes.number,
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

AdventureDelta.defaultProps = {
    starting: 0,
    earned: 0,
    className: '',
    disabled: false,
};

export default AdventureDelta;
