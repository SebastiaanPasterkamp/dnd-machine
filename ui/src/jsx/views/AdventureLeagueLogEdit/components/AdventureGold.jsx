import React from 'react';
import PropTypes from 'prop-types';
import { mergeWith } from 'lodash/fp';

import Coinage from '../../../components/Coinage.jsx';
import ControlGroup from '../../../components/ControlGroup.jsx';
import CostEditor from '../../../components/CostEditor.jsx';
import Panel from '../../../components/Panel.jsx';

export class AdventureGold extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.state = {};
    }

    onChange = (earned) => {
        const { starting, disabled, setState } = this.props;
        if (disabled) {
            return;
        }

        const total = mergeWith(
            (dst, src) => ((dst || 0) + src),
            starting,
            earned,
        );
        this.setState(
            { total },
            () => setState({ starting, earned, total })
        );
    }

    render() {
        const {
            starting, earned, total, className, label, disabled,
        } = this.props;

        return (
            <Panel
                className={className}
                header={label}
            >
                <ControlGroup label="Starting">
                    <Coinage
                        {...starting}
                        className="nice-form-control"
                        extended="1"
                    />
                </ControlGroup>

                <ControlGroup label="Earned">
                    <CostEditor
                        value={earned}
                        disabled={disabled}
                        setState={this.onChange}
                    />
                </ControlGroup>

                <ControlGroup label="Total">
                    <Coinage
                        {...total}
                        className="nice-form-control"
                        extended="1"
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureGold.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: PropTypes.objectOf(PropTypes.number),
    earned: PropTypes.objectOf(PropTypes.number),
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

AdventureGold.defaultProps = {
    starting: {},
    earned: {},
    className: '',
    disabled: false,
};

export default AdventureGold;
