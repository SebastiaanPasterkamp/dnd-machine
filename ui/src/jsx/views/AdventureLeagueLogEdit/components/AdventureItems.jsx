import React from 'react';
import PropTypes from 'prop-types';
import {
    filter,
    isObject,
    map,
} from 'lodash/fp';

import ControlGroup from '../../../components/ControlGroup';
import InputField from '../../../components/InputField';
import Panel from '../../../components/Panel';
import ListComponent from '../../../components/ListComponent';

export class AdventureItems extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.initalItem = {
            value: '',
        };
        this.onChange = this.onChange.bind(this);
    }

    onChange(items) {
        const { starting, disabled, setState } = this.props;
        if (disabled) {
            return;
        }
        const earned = filter(item => item)(
            map(
                item => ( isObject(item) ? item.value : item )
            )(items)
        );
        const total = starting + earned.length;
        setState({ starting, earned, total });
    }

    render() {
        const {
            starting, earned, total, className, label, disabled,
        } = this.props;

        return (
            <Panel
                className={className}
                header={label}
            >
                <ControlGroup label="Starting">
                    <InputField
                        placeholder="Starting..."
                        disabled={true}
                        type="number"
                        value={starting}
                    />
                </ControlGroup>

                <ControlGroup label="Obtained">
                    <ListComponent
                        newItem='auto'
                        list={map( value => ({ value }) )(earned)}
                        component={InputField}
                        initialItem={this.initalItem}
                        disabled={disabled}
                        setState={!disabled
                            ? this.onChange
                            : null
                        }
                        componentProps={{ disabled }}
                    />
                </ControlGroup>

                <ControlGroup label="Total">
                    <InputField
                        placeholder="Total..."
                        disabled={true}
                        type="number"
                        value={total}
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureItems.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: PropTypes.number,
    earned: PropTypes.arrayOf( PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
            item: PropTypes.string,
        }),
    ])),
    total: PropTypes.number,
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

AdventureItems.defaultProps = {
    starting: 0,
    earned: [],
    total: 0,
    className: '',
    disabled: false,
};

export default AdventureItems;
