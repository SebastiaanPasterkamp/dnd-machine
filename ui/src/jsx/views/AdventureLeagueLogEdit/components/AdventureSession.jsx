import React from 'react';
import PropTypes from 'prop-types';

import { memoize } from '../../../utils';

import ControlGroup from '../../../components/ControlGroup';
import InputField from '../../../components/InputField';
import Panel from '../../../components/Panel';

export class AdventureSession extends React.PureComponent
{
    constructor(props) {
        super(props);
        this.memoize = memoize.bind(this);
    }

    onChange(field) {
        return this.memoize(
            field,
            (value) => this.props.setState({ [field]: value })
        );
    }

    render() {
        const {
            name, id, date, dm_name, dm_dci, setState,
        } = this.props;

        return (
            <Panel
                className="al-log-edit__session"
                header="Adventure"
            >
                <ControlGroup label="Name">
                    <InputField
                        data-field="name"
                        placeholder="Name..."
                        value={name}
                        setState={this.onChange('name')}
                    />
                </ControlGroup>

                <ControlGroup label="Session #">
                    <InputField
                        data-field="id"
                        placeholder="Session #..."
                        value={id}
                        setState={this.onChange('id')}
                    />
                </ControlGroup>

                <ControlGroup label="Date">
                    <InputField
                        data-field="date"
                        placeholder="Date..."
                        type="date"
                        value={date}
                        setState={this.onChange('date')}
                    />
                </ControlGroup>

                <ControlGroup label="DM Name">
                    <InputField
                        data-field="dm_name"
                        placeholder="DM Name..."
                        value={dm_name}
                        setState={this.onChange('dm_name')}
                    />
                </ControlGroup>

                <ControlGroup label="DM DCI">
                    <InputField
                        data-field="dm_dci"
                        placeholder="DM DCI..."
                        value={dm_dci}
                        setState={this.onChange('dm_dci')}
                    />
                </ControlGroup>
            </Panel>
        );
    }
};

AdventureSession.propTypes = {
    setState: PropTypes.func,
    name: PropTypes.string,
    id: PropTypes.string,
    date: PropTypes.string,
    dm_name: PropTypes.string,
    dm_dci: PropTypes.string,
};

AdventureSession.defaultProps = {
    name: '',
    id: '',
    date: '',
    dm_name: '',
    dm_dci: '',
};

export default AdventureSession;
