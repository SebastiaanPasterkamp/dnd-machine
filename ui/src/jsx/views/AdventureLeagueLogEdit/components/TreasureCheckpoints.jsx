import React from 'react';
import PropTypes from 'prop-types';
import {
    assign,
    indexOf,
    map,
    reduce,
} from 'lodash/fp';

import { memoize } from '../../../utils';

import CalculatorInputField from '../../../components/CalculatorInputField.jsx';
import ControlGroup from '../../../components/ControlGroup.jsx';
import InputField from '../../../components/InputField.jsx';
import Panel from '../../../components/Panel.jsx';

import { tierType, tierDefault } from '../extraProps';

export class TreasureCheckpoints extends React.PureComponent
{
    tiers = [ 'one', 'two', 'three', 'four' ];

    constructor(props) {
        super(props);
        this.memoize = memoize.bind(this);
    }

    onChange = (tier) => this.memoize(tier, (value) => {
        const {
            disabled,
            starting,
            earned: oldEarned,
            total: oldTotal,
            setState,
        } = this.props;
        if (disabled) {
            return;
        }
        let earned = assign({}, oldEarned);
        earned = assign(earned, {
            [tier]: value,
        });
        let total = assign({}, oldTotal);
        total = assign(total, {
            [tier]: (starting[tier] || 0) + value,
        });

        setState({ starting, earned, total });
    })

    render() {
        const {
            starting, earned, total, className, label, disabled,
            currentTier,
        } = this.props;

        return (
            <Panel
                className={className}
                header={label}
            >
            {map(tier => (
                <ControlGroup
                    key={tier}
                    className={tier === currentTier
                        ? 'current'
                        : undefined
                    }
                    labels={[
                        tier === currentTier
                            ? 'Current tier'
                            : `Tier ${tier}`,
                        "Â±",
                        "="
                    ]}
                >
                    <InputField
                        disabled={true}
                        type="number"
                        value={starting[tier]}
                    />
                    <CalculatorInputField
                        placeholder="Earned..."
                        value={earned[tier]}
                        disabled={disabled}
                        setState={this.onChange(tier)}
                        maxValue={currentTier === tier
                            ? undefined
                            : 0
                        }
                        minValue={(total[tier] || 0) * -1}
                    />
                    <InputField
                        placeholder="Total..."
                        disabled={true}
                        type="number"
                        value={total[tier]}
                    />
                </ControlGroup>
            ))(this.tiers)}
            </Panel>
        );
    }
};

TreasureCheckpoints.propTypes = {
    setState: PropTypes.func,
    label: PropTypes.string.isRequired,
    starting: tierType,
    earned: tierType,
    total: tierType,
    currentTier: PropTypes.oneOf([
        'one', 'two', 'three', 'four',
    ]),
    className: PropTypes.string,
    disabled: PropTypes.bool,
};

TreasureCheckpoints.defaultProps = {
    starting: tierDefault,
    earned: tierDefault,
    total: tierDefault,
    currentTier: 'one',
    className: '',
    disabled: false,
};

export default TreasureCheckpoints;