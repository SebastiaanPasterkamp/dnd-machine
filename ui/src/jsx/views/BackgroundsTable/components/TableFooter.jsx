import React from 'react';

import BackgroundLinks from '../../../components/BackgroundLinks';

export const TableFooter = function() {
    return (
        <tr>
            <td colSpan={2}>
                <BackgroundLinks
                    altStyle={true}
                />
            </td>
        </tr>
    );
};

export default TableFooter;
