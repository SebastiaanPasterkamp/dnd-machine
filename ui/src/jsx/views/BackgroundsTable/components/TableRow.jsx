import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import BackgroundLinks from '../../../components/BackgroundLinks';

export const TableRow = function({ id, name, description }) {
    return (
        <tr data-name={id}>
            <th>
                { name }
                <BackgroundLinks
                    altStyle={true}
                    id={id}
                />
            </th>
            <td>
                <MDReactComponent
                    text={description || ''}
                />
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
};

TableRow.defaultProps = {
    description: '',
};

export default TableRow;
