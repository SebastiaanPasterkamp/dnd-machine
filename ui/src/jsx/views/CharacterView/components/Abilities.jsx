import React from 'react';
import PropTypes from 'prop-types';
import {
    entries,
    flow,
    map,
} from 'lodash/fp';
import MDReactComponent from 'markdown-react-js';
import {sprintf} from 'sprintf-js';

import Panel from '../../../components/Panel.jsx';

const Abilities = function({ abilities }) {
    return (
        <Panel
            key="abilities"
            className="character-view__abilities info"
            header="Abilities"
        >
            <ul>
            {flow(entries, map(([label, data]) => (
                <li key={label}>
                    <strong>{label}</strong>
                    <MDReactComponent
                        text={sprintf(
                            data.description,
                            data
                        )}
                    />
                </li>
            )))(abilities)}
            </ul>
        </Panel>
    );
};

Abilities.propTypes = {
    abilities: PropTypes.objectOf(
        PropTypes.shape({
            description: PropTypes.string.isRequired,
        })
    ),
};

Abilities.defaultProps = {
    abilities: {},
};

export default Abilities;
