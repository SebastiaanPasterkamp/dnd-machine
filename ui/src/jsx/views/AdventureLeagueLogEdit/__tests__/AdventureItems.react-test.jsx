import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { AdventureItems } from '../components/AdventureItems';

describe('AdventureItems', () => {
    const fullProps = {
        label: "Example Items",
        className: "example-items",
        starting: 0,
        earned: ["Health Potion"],
        total: 1,
    };

    it('should render with minimum props', () => {
        const setState = jest.fn();

        const tree = renderer.create(
            <AdventureItems
                setState={setState}
                label={fullProps.label}
            />
        );

        expect(tree).toMatchSnapshot();
    });

    it('should render with full props', () => {
        const setState = jest.fn();

        const tree = renderer.create(
            <AdventureItems
                setState={setState}
                {...fullProps}
            />
        );

        expect(tree).toMatchSnapshot();
    });

    it('should add an item', () => {
        const setState = jest.fn();

        const wrapper = mount(
            <AdventureItems
                setState={setState}
                label={fullProps.label}
            />
        );

        wrapper
            .find('input[disabled=false]')
            .simulate('change', {target: {
                value: "Health Potion",
            }});

        expect(setState).toBeCalledWith({
            starting: 0,
            earned: ["Health Potion"],
            total: 1,
        });
    });

    it('should remove an item', () => {
        const setState = jest.fn();

        const wrapper = mount(
            <AdventureItems
                setState={setState}
                {...fullProps}
            />
        );

        wrapper
            .find('input[disabled=false]')
            .at(0)
            .simulate('change', {target: {
                value: "",
            }});

        expect(setState).toBeCalledWith({
            starting: 0,
            earned: [],
            total: 0,
        });
    });

    it('should remove an empty item', () => {
        const setState = jest.fn();

        const wrapper = mount(
            <AdventureItems
                setState={setState}
                {...fullProps}
            />
        );

        wrapper
            .find('button[name="del"]')
            .simulate('click');

        expect(setState).toBeCalledWith({
            starting: 0,
            earned: [],
            total: 0,
        });
    });
});