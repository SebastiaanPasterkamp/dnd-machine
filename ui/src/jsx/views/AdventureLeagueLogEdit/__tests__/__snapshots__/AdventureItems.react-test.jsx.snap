// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdventureItems should render with full props 1`] = `
<div
  className="nice-panel example-items"
  id={undefined}
>
  <div
    className="nice-panel-heading"
  >
    Example Items
  </div>
  <div
    className="nice-panel-content"
  >
    <div
      className="nice-control-group"
    >
      <span
        className="nice-input-addon"
      >
        Starting
      </span>
      <input
        className="nice-form-control"
        disabled={true}
        onChange={[Function]}
        onKeyPress={null}
        placeholder="Starting..."
        type="number"
        value={0}
      />
    </div>
    <div
      className="nice-control-group"
    >
      <span
        className="nice-input-addon"
      >
        Obtained
      </span>
      <ul
        className="list-component"
      >
        <li>
          <input
            className="nice-form-control"
            disabled={false}
            onChange={[Function]}
            onKeyPress={null}
            placeholder={undefined}
            type={undefined}
            value="Health Potion"
          />
          <button
            className="nice-btn list-component__delete warning icon fa-minus"
            name="del"
            onClick={[Function]}
          >
            â€‹
          </button>
        </li>
        <li>
          <input
            className="nice-form-control"
            disabled={false}
            onChange={[Function]}
            onKeyPress={null}
            placeholder={undefined}
            type={undefined}
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      className="nice-control-group"
    >
      <span
        className="nice-input-addon"
      >
        Total
      </span>
      <input
        className="nice-form-control"
        disabled={true}
        onChange={[Function]}
        onKeyPress={null}
        placeholder="Total..."
        type="number"
        value={1}
      />
    </div>
    <div
      className="clearfix"
    />
  </div>
</div>
`;

exports[`AdventureItems should render with minimum props 1`] = `
<div
  className="nice-panel"
  id={undefined}
>
  <div
    className="nice-panel-heading"
  >
    Example Items
  </div>
  <div
    className="nice-panel-content"
  >
    <div
      className="nice-control-group"
    >
      <span
        className="nice-input-addon"
      >
        Starting
      </span>
      <input
        className="nice-form-control"
        disabled={true}
        onChange={[Function]}
        onKeyPress={null}
        placeholder="Starting..."
        type="number"
        value={0}
      />
    </div>
    <div
      className="nice-control-group"
    >
      <span
        className="nice-input-addon"
      >
        Obtained
      </span>
      <ul
        className="list-component"
      >
        <li>
          <input
            className="nice-form-control"
            disabled={false}
            onChange={[Function]}
            onKeyPress={null}
            placeholder={undefined}
            type={undefined}
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      className="nice-control-group"
    >
      <span
        className="nice-input-addon"
      >
        Total
      </span>
      <input
        className="nice-form-control"
        disabled={true}
        onChange={[Function]}
        onKeyPress={null}
        placeholder="Total..."
        type="number"
        value={0}
      />
    </div>
    <div
      className="clearfix"
    />
  </div>
</div>
`;
